### PR TITLE
[Feature] support structure output for model runner v2

### PR DESCRIPTION
**Pull request by Meihan-chen**
### What this PR does / why we need it?

This PR adds Ascend NPU support for structured outputs in model runner v2.

Upstream `StructuredOutputsWorker.apply_grammar_bitmask` cannot be used directly on Ascend because the original Triton kernel with `BLOCK_SIZE=8192` may exceed UB capacity.  Reducing `BLOCK_SIZE` avoids UB overflow, but increases the number of work-groups and may lead to unstable scheduling behavior on NPU.

To address this, this PR adds an Ascend-specific `_apply_grammar_bitmask_kernel` implementation that:
- introduces `BLOCK_SIZE_SUB=1024` tiling inside each block to reduce UB usage
- applies the grammar bitmask to logits with an NPU-compatible Triton kernel
- patches `_apply_grammar_bitmask_kernel` for model runner v2

### How was this patch tested?

- vLLM version: v0.19.0
- vLLM main: https://github.com/vllm-project/vllm/commit/6f786f2c506cb07f4566771fdc62e640e2c4a176

### CODE REVIEW

File: vllm_ascend/worker/v2/structured_outputs.py [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+
+    # Construct bitmask -> logits mapping
+    mapping: list[int] = []
+    req_ids = input_batch.req_ids
+    cu_num_logits = input_batch.cu_num_logits_np.tolist()
+    req_id_to_idx = {req_id: i for i, req_id in enumerate(req_ids)}
+    for grammar_req_id in grammar_req_ids:
+        req_idx = req_id_to_idx[grammar_req_id]
+        logits_start_idx = cu_num_logits[req_idx]
+        logits_end_idx = cu_num_logits[req_idx + 1]
```
Comment (Sparkheart) [inline vllm_ascend/worker/v2/structured_outputs.py]: req_idx: Check whether the value exceeds the length of cu_num_logits. If yes, modify it. If no, ignore it.

File: vllm_ascend/worker/v2/structured_outputs.py [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+    # Asynchronously copy the mapping to NPU.
+    with torch.cuda.stream(self.copy_stream):
+        logits_indices_cpu = torch.tensor(mapping, dtype=torch.int32, device="cpu", pin_memory=True)
+        logits_indices = self.logits_indices[: len(mapping)].copy_(logits_indices_cpu, non_blocking=True)
+    # ensure copies finish before kernel launch
+    current_stream = torch.cuda.current_stream()
+    current_stream.wait_stream(self.copy_stream)
+
+    num_masks = bitmask.shape[0]
+    assert num_masks == len(mapping), f"num_masks={num_masks} != len(mapping)={len(mapping)}"
```
Comment (Sparkheart) [inline vllm_ascend/worker/v2/structured_outputs.py]: Could the position of the assertion be considered moving forward a bit?

Comment (Meihan-chen) [reply vllm_ascend/worker/v2/structured_outputs.py]: Thanks, moved the `num_masks == len(mapping)` assertion earlier

File: vllm_ascend/worker/v2/structured_outputs.py (hunk 1 above)
Comment (Meihan-chen) [reply vllm_ascend/worker/v2/structured_outputs.py]: Thanks, `cu_num_logits_np` is built as a prefix-sum array with length `len(req_ids) + 1`, so `req_idx + 1` should be in bounds here. I think no code change is needed for this part.

File: vllm_ascend/worker/v2/structured_outputs.py [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+
+    # Asynchronously copy the mapping to NPU.
+    with torch.cuda.stream(self.copy_stream):
+        logits_indices_cpu = torch.tensor(mapping, dtype=torch.int32, device="cpu", pin_memory=True)
+        logits_indices = self.logits_indices[: len(mapping)].copy_(logits_indices_cpu, non_blocking=True)
+    # ensure copies finish before kernel launch
+    current_stream = torch.cuda.current_stream()
```
Comment (MengqingCao) [inline vllm_ascend/worker/v2/structured_outputs.py]: plz use `torch.npu.current_stream()` instead

**Comment by Ronald1995** [comment]
please add e2e test for this feature.

File: vllm_ascend/worker/v2/structured_outputs.py [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+
+import torch
+from vllm.triton_utils import tl, triton
+from vllm.v1.worker.gpu.buffer_utils import async_copy_to_gpu
+
+_NPU_BLOCK_SIZE = 8192
+_NPU_BLOCK_SIZE_SUB = 1024
+
+
+def apply_grammar_bitmask(self, logits, input_batch, grammar_req_ids, grammar_bitmask):
```
Comment (realliujiaxu) [inline vllm_ascend/worker/v2/structured_outputs.py]: patch `_apply_grammar_bitmask_kernel` rather than `apply_grammar_bitmask`

**Review by realliujiaxu — CHANGES_REQUESTED** [review]

File: vllm_ascend/worker/v2/structured_outputs.py (hunk 3 above)
Comment (Meihan-chen) [reply vllm_ascend/worker/v2/structured_outputs.py]: thanks，all torch.cuda have changed

File: vllm_ascend/worker/v2/structured_outputs.py (hunk 4 above)
Comment (Meihan-chen) [reply vllm_ascend/worker/v2/structured_outputs.py]: Thanks . I updated the patch to override `_apply_grammar_bitmask_kernel` instead of replacing the whole `apply_grammar_bitmask` method.

**Comment by Meihan-chen** [comment]
> please add e2e test for this feature.

thx，i have added a test

**Comment by Meihan-chen** [comment]
@MengqingCao  @realliujiaxu  Could you please take a look?

File: tests/e2e/singlecard/test_guided_decoding.py [hunk 5]
```diff
@@ -26,12 +28,23 @@
...
 MODEL_NAME = "Qwen/Qwen3-0.6B"
 
 GuidedDecodingBackend = ["xgrammar", "guidance", "outlines"]
 
 
+@pytest.fixture(params=[False, True], ids=["v1", "v2"])
+def model_runner_env(request):
+    use_v2_model_runner = request.param
+    if use_v2_model_runner and vllm_version_is("0.20.1"):
+        pytest.skip("No need to support v2 model runner for vLLM tag version.")
```
Comment (MengqingCao) [inline tests/e2e/singlecard/test_guided_decoding.py:42]: why we doesn't support 0.20.1 here?

File: vllm_ascend/worker/v2/structured_outputs.py [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+# Upstream cannot be used directly on Ascend NPU: `BLOCK_SIZE=8192` overflows
+# UB, while a smaller `BLOCK_SIZE` makes the grid unstable. We therefore keep
+# `BLOCK_SIZE=8192` and split each block with `BLOCK_SIZE_SUB=1024`.
+#
+#
+
+from vllm.triton_utils import tl, triton
+
+
+# Adapted from
```
Comment (MengqingCao) [inline vllm_ascend/worker/v2/structured_outputs.py:30]: Let's add a todo here for performance optimization

File: tests/e2e/singlecard/test_guided_decoding.py (hunk 5 above)
Comment (Meihan-chen) [reply tests/e2e/singlecard/test_guided_decoding.py:42]: The current strategy for model_runner v2 is to only adapt to the main branch.

File: vllm_ascend/worker/v2/structured_outputs.py (hunk 6 above)
Comment (Meihan-chen) [reply vllm_ascend/worker/v2/structured_outputs.py:30]: done

**Review by realliujiaxu — APPROVED** [review]

**Comment by realliujiaxu** [comment]
LGTM

**Comment by Meihan-chen** [comment]
@MengqingCao CI is passing now, could you please merge this PR？

**Review by MengqingCao — APPROVED** [review]
